gateway/sensor.go

```diff
@@ -68,22 +67,88 @@ func init() {
 
 // Validate ensures all parts of the config are valid.
 func (conf *Config) Validate(path string) ([]string, error) {
+	for _, d := range conf.Devices {
+		if d.DecoderPath == "" {
+			return nil, resource.NewConfigValidationError(path,
+				errors.New("decoder path is required"))
+		}
+		switch d.JoinType {
+		case "ABP":
+			return d.validateABPAttributes(path)
+		case "OTAA", "":
+			return d.validateOTAAAttributes(path)
+		default:
+			return nil, resource.NewConfigValidationError(path,
+				errors.New("join type is OTAA or ABP - Default is OTAA"))
+		}
+
+	}
+	return nil, nil
+}
+
+func (conf *DeviceConfig) validateOTAAAttributes(path string) ([]string, error) {
+	if conf.DevEUI == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("dev EUI is required for OTAA join type"))
+	}
+	if len(conf.DevEUI) != 16 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("dev EUI must be 8 bytes"))
+	}
+	if conf.AppKey == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app key is required for OTAA join type"))
+	}
+	if len(conf.AppKey) != 32 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app key must be 16 bytes"))
+	}
 	return nil, nil
 }
 
+func (conf *DeviceConfig) validateABPAttributes(path string) ([]string, error) {
+	if conf.AppSKey == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app session key is required for ABP join type"))
+	}
+	if len(conf.AppSKey) != 32 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app session key must be 16 bytes"))
+	}
+	if conf.NwkSKey == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("network session key is required for ABP join type"))
+	}
+	if len(conf.NwkSKey) != 32 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("network session key must be 16 bytes"))
+	}
+	if conf.DevAddr == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("device address is required for ABP join type"))
+	}
+	if len(conf.DevAddr) != 8 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("device address must be 4 bytes"))
+	}
+
+	return nil, nil
+
+}
+
 type Gateway struct {
 	resource.Named
 	resource.AlwaysRebuild
 	logger logging.Logger
 
 	workers *utils.StoppableWorkers
+	mu      sync.Mutex
 
-	lastReadings map[string]interface{} // map devices to their last reading
+	lastReadings map[string]interface{} // map of devices to readings
+	readingsMu   sync.Mutex
```

Comment:
keep an eye on the two mutexes as you code more; if it makes sense to move the readings mutex and last reading into a struct that has a separate mutex hold to write and overwrite, we should do that early in this module's life.

Places we care about double locking the struct are the constructor + `Reconfigure`, `Close`, and `Readings` being spammed by multiple clients + data capture.

Reply:
the `mu` mutex is to avoid sending on a gateway channel at the same, but I think that we should be able to get the readings at the same time as a send hence the two mutexes. But maybe just one would be sufficient.